src/core/lib/promise/loop.h

```diff
@@ -0,0 +1,95 @@
+// Copyright 2021 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GRPC_CORE_LIB_PROMISE_LOOP_H
+#define GRPC_CORE_LIB_PROMISE_LOOP_H
+
+#include <grpc/impl/codegen/port_platform.h>
+
+#include "absl/types/variant.h"
+#include "src/core/lib/promise/detail/promise_factory.h"
+
+namespace grpc_core {
+
+// Special type - signals to loop to take another iteration, instead of
+// finishing
+struct Continue {};
+
+// Result of polling a loop promise - either Continue looping, or return a value
+// T
+template <typename T>
+using LoopCtl = absl::variant<Continue, T>;
+
+namespace promise_detail {
+
+template <typename T>
+struct LoopTraits;
+template <typename T>
+struct LoopTraits<LoopCtl<T>> {
+  using Result = T;
+};
+
+template <typename F>
+class Loop {
+ private:
+  using Factory = promise_detail::PromiseFactory<void, F>;
+  using Promise = decltype(std::declval<Factory>().Repeated());
+  using PromiseResult = typename Promise::Result;
```

Comment:
nit: Does this offer anything beyond syntax sugar? If this is here solely to avoid typing `::`, consider removing `PromiseResult`.

Reply:
This one's tough.. we only use it once, a few lines down. It would have to be `typename Promise::Result` there too, and probably spill that type expression over multiple lines, and I'm unsure which is the greater evil.

Reply:
Gotcha. I'm fine to leave it as is, nit retracted.